src/onnx/parse_softmaxcrossentropyloss.cpp

```diff
@@ -0,0 +1,308 @@
+
+        // TODO: Update this operator when we get bfloat16 support
+        std::set<migraphx::shape::type_t> supported_score_weight_types = {
+            migraphx::shape::half_type, migraphx::shape::float_type, migraphx::shape::double_type};
+        if(not contains(supported_score_weight_types, scores_shape.type()))
+        {
+            MIGRAPHX_THROW(
+                "softmaxcrossentropyloss: Score must be either half, float, or double type");
+        }
+
+        if(scores_shape.lens()[0] != label_shape.lens()[0])
+        {
+            MIGRAPHX_THROW(
+                "softmaxcrossentropyloss: Score and Labels must identical batch size inputs");
+        }
+
+        if((scores_shape.ndim() - 1) != label_shape.ndim())
+        {
+            MIGRAPHX_THROW(
+                "softmaxcrossentropyloss: Score and Labels must contain identical K-Dimensions");
+        }
+
+        bool is_k_dim  = (scores_shape.ndim() > 3);
+        size_t batch_size = scores_shape.lens().at(0);
+        size_t class_size = scores_shape.lens().at(1);
+
+        auto loss_tensor = info.add_instruction(
+            migraphx::make_op("convert", {{"target_type", scores_shape.type()}}), labels);
+
+        // Default weights will always be 1
+        auto weights = info.add_literal(
+            migraphx::literal(migraphx::shape(scores_shape.type(), {1}, {0}), {1}));
+        weights = info.add_instruction(
+            migraphx::make_op("multibroadcast", {{"out_lens", {class_size}}}), weights);
+
+        // Get optional input weights (Used for mean reduction)
+        bool has_weights = (args.size() > 2);
+        if(has_weights)
+        {
+            weights = args.at(2);
+            auto weights_shape = weights->get_shape();
+
+            if(weights_shape.lens()[0] != scores_shape.lens()[1])
+            {
+                MIGRAPHX_THROW("softmaxcrossentropyloss: Invalid weight vector shape. Weight must "
+                               "contain weight for each class");
+            }
+
+            if(not contains(supported_score_weight_types, weights_shape.type()))
+            {
+                MIGRAPHX_THROW(
+                    "softmaxcrossentropyloss: weight must be either half, float, or double type");
+            }
+
+            if(weights_shape.type() != scores_shape.type())
+            {
+                MIGRAPHX_THROW(
+                    "softmaxcrossentropyloss: Weight and Scores inputs must be the same type");
+            }
+        }
+
+        // adjust weights based on ignore index is that's set
+        if(has_ignore_index)
+        {
+            auto weights_shape = weights->get_shape();
+            std::vector<float> zero_val_vect(weights_shape.elements(), 0);
+            auto zero_val = info.add_literal(migraphx::literal(weights_shape, zero_val_vect));
+            weights       = info.add_instruction(
+                migraphx::make_op("scatter_none", {{"axis", 0}}), weights, ignore_index, zero_val);
+        }
```

Comment:
I do not think this directly corresponds to onnx reference code here 
https://github.com/onnx/onnx/blob/b90e252da11dea9bdc191d6b9b8d01511ef3e3bd/onnx/reference/ops/op_softmax_cross_entropy_loss.py#L33 

Weights have shape `[C]` where C is number of classes. 
From what i understand each element in weight corresponds to weight for that class index's weight. 

e.g. let's say weights  = [3, 2, 1] then weight for class_index = 0 would be 3, weight for class_index = 1 would be 2 weight for class_index = 2 would be 1. 

That is why onnx reference code is first doing `np.take` which would map weights to `target`. 

e.g. if lets' say target_indices = [1, 1, 2] then actual weights to use would be [2, 2, 1]. 

And then ignore_index is ignore_index for the value inside target_indices vector. 

So let's say if `ignore_index = 2` then target_indices would become [1, 1, ignore] and then actual weights to use would become `[2, 2, 0]` 

You are scattering zero values inside the weight matrix directly which is not the same as above and onnx's ref.

Reply:
Ah the intent here was to avoid the use of where here and needing to create a literal of masks to set the zero value for the weight index. At the end I  always do the multiply with the weights so for the default case , everything in weighs is all 1s, 

Scatter_none should do the update of the zero at the value specified by ignore_index which we read in, thus making the weight zero the final result will be zero when scaled out and any loss will be cancelled to zero. I've tried it

I've got this in a test case thats passing based on the input gold data I used and had ignore_index = 2, and I used weights that were different to ensure I'm getting the right weight zeroed out. The weight vector is always a 1d vector in the spec bound by classes so your initial assertion is correct. I didn't see an equivalent with take so I tried to handle it with this. The verify test for this I added is 

`softmaxcrossentropyloss_2d_no_reduction_uneven_weighted_ignore_index_test`

Here's the output up to the scatter with MIGRAPHX_TRACE_EVAL=2
```
[   RUN    ] softmaxcrossentropyloss_2d_no_reduction_uneven_weighted_ignore_index_test
Run instruction: @0 = @literal{0, 0, 0, 0} -> float_type, {4}, {1}, target_id=0
Time: 0.01746ms, 0.018961ms
Run instruction: @1 = @literal{2} -> int64_type, {1}, {0}, target_id=0
Time: 0.00136ms, 0.00146ms
Run instruction: 2 = @param:2 -> float_type, {4}, {1}, target_id=0
Time: 0.01504ms, 0.01515ms
Run instruction: 1 = @param:1 -> int32_type, {4}, {1}, target_id=0
Time: 0.00304ms, 0.00313ms
Run instruction: 0 = @param:0 -> float_type, {4, 4}, {4, 1}, target_id=0
Time: 0.00144ms, 0.00154ms
Run instruction: @5 = ref::convert[target_type=2](1) -> float_type, {4}, {1}, target_id=0
Time: 0.08942ms, 0.08957ms
Output has normal, zero
Output: 0, 3, 1, 2
Min value: 0, Max value: 3, Mean: 1.5, StdDev: 1.11803
Run instruction: @6 = ref::scatter_none[axis=0](2,@1,@0) -> float_type, {4}, {1}, target_id=0
Time: 0.09062ms, 0.0909ms
Output has zero, normal
Output: 1, 0.5, 0, 3
Min value: 0, Max value: 3, Mean: 1.125, StdDev: 1.1388
```

Reply:
I see what you are trying to do  
It would be better to keep all weights/ignore index related calculation together in the code. 
Weights/ignore indices are not used untill later.

Reply:
Agreed. I need to use ignore index as soon as I read in a valid input score as that'll give me the max class size. I've modified this to add better handling of out of bounds ignore_index and cleaned it up a bit. I can move things around.